middleware/apimiddleware.go

```diff
@@ -414,21 +407,61 @@ func GetAddressCtx(r *http.Request, activeNetParams *chaincfg.Params, maxAddrs i
 		return false
 	}
 
-	var addrStrs []string
+	// Allocate as if all addresses are unique.
+	addrStrs := make([]string, 0, len(addressStrs))
+	for _, addrStr := range addressStrs {
+		if strInSlice(addrStrs, addrStr) {
+			continue
+		}
+		addrStrs = append(addrStrs, addrStr)
+	}
+
 	for _, addrStr := range addressStrs {
 		_, err := dcrutil.DecodeAddress(addrStr, activeNetParams)
 		if err != nil {
 			return nil, fmt.Errorf("invalid address '%v' for this network: %v",
 				addrStr, err)
 		}
-		if strInSlice(addrStrs, addrStr) {
-			continue
-		}
-		addrStrs = append(addrStrs, addrStr)
 	}
 	return addrStrs, nil
 }
 
+// GetAddressRawCtx returns a slice of addresses parsed from the {address} URL
+// parameter. Multiple comma-delimited address strings can be specified.
+func GetAddressRawCtx(r *http.Request, activeNetParams *chaincfg.Params, maxAddrs int) ([]dcrutil.Address, error) {
+	addressStrs, err := addressStrings(r, maxAddrs)
+	if err != nil {
+		return nil, err
+	}
+	addresses := make([]dcrutil.Address, 0, len(addressStrs))
+	for _, addrStr := range addressStrs {
+		addr, err := dcrutil.DecodeAddress(addrStr, activeNetParams)
+		if err != nil {
+			return nil, fmt.Errorf("invalid address '%v' for this network: %v",
+				addrStr, err)
+		}
+		addresses = append(addresses, addr)
+	}
+	return addresses, nil
+}
+
+// addressStrings retrieves the CtxAddress data from the request context. If not
+// set, an error is generated. The CtxAddress string data may be a
+// comma-separated list of addresses, subject to the maxAddrs limit.
+func addressStrings(r *http.Request, maxAddrs int) ([]string, error) {
+	addressStr, ok := r.Context().Value(CtxAddress).(string)
+	if !ok || len(addressStr) == 0 {
+		apiLog.Trace("address not set")
+		return nil, fmt.Errorf("address not set")
+	}
+	addressStrs := strings.Split(addressStr, ",")
+	if len(addressStrs) > maxAddrs {
+		return nil, fmt.Errorf("maximum of %d addresses allowed", maxAddrs)
+	}
+
+	return addressStrs, nil
+}
+
```

Comment:
Nice.  Good validation.  Although I'm thinking it could be even better to have this further up in the request handling chain, where `CtxAddress` is set (`middleware.AddressPathCtx`) to preempt the http handlers entirely if any of the addresses are invalid.  If this opens a can of worms or becomes a major overhaul of all the `/addr/...` endpoints, let's skip it though... just a thought.
It's basically the idea applied in the [new indent middleware](https://github.com/decred/dcrdata/pull/1713/files#diff-92898af3f3b4aa2e3e620614c71dbf9aR542-R551) - validate the url bits and call `next.ServerHTTP` otherwise `http.Error` so that no other middlewares, subrouters, or any parts of the routed handler actually have to run if validation fails.